community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/KeysFunction.scala

```diff
@@ -21,46 +21,40 @@ package org.neo4j.cypher.internal.compiler.v2_2.commands.expressions
 
 import org.neo4j.cypher.internal.compiler.v2_2._
 import org.neo4j.cypher.internal.compiler.v2_2.executionplan.Effects
-import pipes.QueryState
-import symbols._
-import org.neo4j.cypher.internal.compiler.v2_2.spi.QueryContext
-import org.neo4j.graphdb.{Relationship, Node}
+import org.neo4j.cypher.internal.compiler.v2_2.helpers.IsMap
+import org.neo4j.cypher.internal.compiler.v2_2.pipes.QueryState
+import org.neo4j.cypher.internal.compiler.v2_2.symbols._
+import org.neo4j.graphdb.{Node, Relationship}
 
-case class KeysFunction(nodeOrRelationshipExpr: Expression) extends NullInNullOutExpression(nodeOrRelationshipExpr) {
+case class KeysFunction(expr: Expression) extends NullInNullOutExpression(expr) {
 
-  override def compute(value: Any, m: ExecutionContext)(implicit state: QueryState) = value match {
-    case node: Node        =>
-                              val queryCtx: QueryContext = state.query
-                              queryCtx.getPropertiesForNode(node.getId).map { case (v) =>
-                                queryCtx.getPropertyKeyName(v.toInt)
-                              }.toList
-    case rel: Relationship =>
-                              val queryCtx: QueryContext = state.query
-                              queryCtx.getPropertiesForRelationship(rel.getId).map { case (v) =>
-                                queryCtx.getPropertyKeyName(v.toInt)
-                              }.toList
-    case x => throw new CypherTypeException("Expected `%s` to be a node or relationship, but it was ``".format(nodeOrRelationshipExpr, x.getClass.getSimpleName))
+  override def compute(value: Any, ctx: ExecutionContext)(implicit state: QueryState) = value match {
+    case IsMap(map) => map(state.query).keys.toList
```

Comment:
Nothing like to have a good knowledge about the core. ;) less code, less efforts, less issues ...it's cool!